File: host/guest_test.go
package host

import (
	"os"
	"testing"
	"time"

	"github.com/ayonli/goext"
	"github.com/ayonli/ngrpc/config"
	"github.com/ayonli/ngrpc/util"
	"github.com/stretchr/testify/assert"
)

func TestNewGuest(t *testing.T) {
	app := config.App{
		Name: "example-server",
		Url:  "grpc://localhost:4000",
	}
	handleStop := func(msgId string) {}
	guest := NewGuest(app, handleStop)

	assert.Equal(t, app.Name, guest.AppName)
	assert.Equal(t, app.Url, guest.AppUrl)
	assert.Equal(t, 0, guest.state)
	assert.NotNil(t, guest.handleStopCommand)
}

func TestGuest_JoinAndLeave(t *testing.T) {
	goext.Ok(0, util.CopyFile("../ngrpc.json", "ngrpc.json"))
	goext.Ok(0, util.CopyFile("../tsconfig.json", "tsconfig.json"))
	defer os.Remove("ngrpc.json")
	defer os.Remove("tsconfig.json")

	cfg := goext.Ok(config.LoadConfig())
	host := NewHost(cfg, false)
	goext.Ok(0, host.Start(false))
	defer host.Stop()

	c := make(chan string)
	guest := NewGuest(config.App{
		Name: "example-server",
		Url:  "grpc://localhost:4000",
	}, func(msgId string) {
		c <- msgId
	})
	guest.Join()

	assert.Equal(t, 1, guest.state)
	assert.Equal(t, 1, len(host.clients))

	guest.Leave("app [example-server] stopped", "")

	time.Sleep(time.Millisecond * 10) // wait a while for the host to close the connection
	assert.Equal(t, 0, guest.state)
	assert.Equal(t, 0, len(host.clients))
}

func TestGuest_JoinRedundantSocketFile(t *testing.T) {
	sockFile, _ := GetSocketPath()
	os.WriteFile(sockFile, []byte{}, 0644)

	assert.True(t, util.Exists(sockFile))

	guest := NewGuest(config.App{
		Name: "example-server",
		Url:  "grpc://localhost:4000",
	}, func(msgId string) {})
	err := guest.connect()

	assert.NotNil(t, err)
	assert.False(t, util.Exists(sockFile))
}

File: pm/guest.test.ts
import "@ayonli/jsext/function";
import { test } from "mocha";
import * as assert from "assert";
import * as path from "path";
import * as fs from "fs";
import { App } from "../app";
import { exists } from "../util";
import { Guest, ControlMessage, encodeMessage, decodeMessage, getSocketPath } from "./guest";

function newMsg(msg: ControlMessage): ControlMessage {
    msg.app ??= "";
    msg.error ??= "";
    msg.fin ??= false;
    msg.msgId ??= "";
    msg.pid ??= 0;
    msg.guests ??= [];
    msg.text ??= "";

    return { ...msg };
}

test("encodeMessage", () => {
    const msg = encodeMessage(newMsg({ cmd: "stop", app: "example-server", msgId: "abc" }));
    assert.strictEqual(msg[msg.length - 1], "\n");
});

test("decodeMessage", () => {
    const msg: ControlMessage = newMsg({ cmd: "stop", app: "example-server", msgId: "abc" });
    const data = encodeMessage(msg);
    let packet = "";
    let buf = data.slice(0, 256);

    const { packet: _packet, messages } = decodeMessage(packet, buf, false);
    packet = _packet;

    assert.strictEqual(messages.length, 1);
    assert.deepStrictEqual(messages[0], msg);
    assert.strictEqual(packet, "");
});

test("decodeMessage overflow", () => {
    const msg: ControlMessage = newMsg({ cmd: "stop", app: "example-server", msgId: "abc" });
    const data = encodeMessage(msg);
    let packet = "";
    let buf = data.slice(0, 64);
    let offset = 0;

    let result = decodeMessage(packet, buf, false);
    let messages = result.messages;
    packet = result.packet;
    offset += 64;

    assert.strictEqual(messages.length, 0);
    assert.strictEqual(packet, buf);

    while (offset < data.length) {
        buf = data.slice(offset, offset + 64);
        const result = decodeMessage(packet, buf, false);
        messages = result.messages;
        packet = result.packet;
        offset += 64;
    }

    assert.strictEqual(messages.length, 1);
    assert.deepStrictEqual(messages[0], msg);
    assert.strictEqual(packet, "");
});

test("decodeMessage EOF", () => {
    const msg: ControlMessage = newMsg({ cmd: "stop", app: "example-server", msgId: "abc" });
    const data = encodeMessage(msg).slice(0, -1);
    let packet = "";
    let buf = data.slice(0, 256);

    let result = decodeMessage(packet, buf, true);
    let messages = result.messages;
    packet = result.packet;

    assert.strictEqual(messages.length, 1);
    assert.deepStrictEqual(messages[0], msg);
    assert.strictEqual(packet, "");
});

test("getSocketPath", () => {
    const cwd = process.cwd();
    const { sockFile, sockPath } = getSocketPath();

    assert.strictEqual(sockFile, path.join(cwd, "ngrpc.sock"));

    if (process.platform === "win32") {
        assert.strictEqual(sockPath, "\\\\.\\pipe\\" + path.join(cwd, "ngrpc.sock"));
    } else {
        assert.strictEqual(sockPath, path.join(cwd, "ngrpc.sock"));
    }
});

test("new Guest", () => {
    const app: App = {
        name: "example-server",
        url: "grpc://localhost:4000",
        services: [],
    };
    const handleStop = () => void 0;
    const handleReload = () => void 0;
    const guest = new Guest(app, { onStopCommand: handleStop, onReloadCommand: handleReload });

    assert.strictEqual(guest.appName, app.name);
    assert.strictEqual(guest.appUrl, app.url);
    assert.strictEqual(guest["handleStopCommand"], handleStop);
    assert.strictEqual(guest["handleReloadCommand"], handleReload);
});

test("Guest join redundant socket file", async () => {
    const { sockFile } = getSocketPath();
    fs.writeFileSync(sockFile, Buffer.from([]), "binary");

    assert.ok(await exists(sockFile));

    const guest = new Guest({
        name: "example-server",
        url: "grpc://localhost:4000",
        services: [],
    }, {
        onStopCommand: () => void 0,
        onReloadCommand: () => void 0,
    });
    const [err] = await Function.try(() => guest["connect"]());

    assert.ok(!!err);
    assert.ok(!(await exists(sockFile)));
});
